### PR TITLE
Fix some pyright issues in parser.py

### PR DESCRIPTION
**Pull request by math-fehr**

### CODE REVIEW

File: xdsl/parser.py [hunk 1]
```diff
@@ -609,7 +608,7 @@ def __init__(self,
                  ctx: MLContext,
                  input: str,
                  name: str = '<unknown>',
-                 allow_unregistered_ops=False):
+                 allow_unregistered_ops: bool = False):
```
Comment (webmiche) [inline xdsl/parser.py:611]: Do you really have to annotated keyword args? This looks extremely weird. Wait, could you use the variable `allow_unregistered_ops` afterward? Please say no.

File: xdsl/parser.py [hunk 2]
```diff
@@ -688,7 +687,7 @@ def _parse_optional_block_label(
         A block label consists of block-id ( `(` block-arg `,` ... `)` )?
         """
         block_id = self.try_parse_block_id()
-        arg_list = list()
+        arg_list = list[tuple[Span, Attribute]]()
```
Comment (webmiche) [inline xdsl/parser.py:690]: What is the difference to `arg_list : List[...] = []`? Does this thing just do type inference instead of explicit annotatios?

**Review by webmiche — APPROVED** [review]

File: xdsl/parser.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/parser.py:611]: yeah the type annotations are required for all arguments

File: xdsl/parser.py (hunk 2 above)
Comment (superlopuh) [reply xdsl/parser.py:690]: it does for local variables

**Review by superlopuh — APPROVED** [review]
🙏

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**